### PR TITLE
fix(NODE-4423): better type support for nested objects in query & update

### PR DESCRIPTION
**Pull request by coyotte508**
#### What is changing?

Better typing for update & query

#### What is the motivation for this change?

```ts
	interface User {
		pet: {
			name: {
				first: string;
				last: string;
			};
		};
	}
	
	// This works when it shouldn't
	const query: mongodb.Filter<User> = { "pet.name": { first: 1 } };
	// This doesn't work when it should
	const update: mongodb.UpdateFilter<User> = { $set: { "pet.name": { first: "a", last: "b" } } };
```

The second error is a regression. This PR fixes both issues

Tickets:

- https://jira.mongodb.org/browse/NODE-4423
- https://jira.mongodb.org/browse/NODE-4450 (duplicate)

### Double check the following

- [x] Ran `npm run check:lint` script
- [x] Self-review completed using the [steps outlined here](https://github.com/mongodb/node-mongodb-native/blob/HEAD/CONTRIBUTING.md#reviewer-guidelines)
- [x] PR title follows the correct format: `<type>(NODE-xxxx)<!>: <description>`
- [x] Changes are covered by tests
- [x] New TODOs have a related JIRA ticket

### CODE REVIEW

File: test/types/community/collection/updateX.test-d.ts [hunk 1]
```diff
@@ -201,9 +206,17 @@ expectAssignable<UpdateFilter<TestModel>>({
...
 expectAssignable<UpdateFilter<TestModel>>({ $set: { stringField: 'a' } });
 expectError(buildUpdateFilter({ $set: { stringField: 123 } }));
 expectAssignable<UpdateFilter<TestModel>>({ $set: { 'subInterfaceField.field2': '2' } });
+expectAssignable<UpdateFilter<TestModel>>({ $set: { 'subInterfaceField.nestedObject.a': '2' } });
+expectAssignable<UpdateFilter<TestModel>>({
+  $set: { 'subInterfaceField.nestedObject': { a: '1', b: '2' } }
+});
+expectError<UpdateFilter<TestModel>>({
+  $set: { 'subInterfaceField.nestedObject': { a: '1' } }
+});
```
Comment (dariakp) [inline test/types/community/collection/updateX.test-d.ts:219]:
```suggestion
expectError<UpdateFilter<TestModel>>({
  $set: { 'subInterfaceField.nestedObject': { a: '1' } }
});
expectError<UpdateFilter<TestModel>>({
  $set: { 'subInterfaceField.nestedObject': { a: 1, 'b': '2' } }
});
```

**Review by dariakp — COMMENTED** [review]
Thanks for the submission! The tests look solid, I'm adding a few clean up items and will revisit once this passes our CI.

File: test/types/community/collection/updateX.test-d.ts [hunk 2]
```diff
@@ -201,9 +206,20 @@ expectAssignable<UpdateFilter<TestModel>>({
...
+expectError<UpdateFilter<TestModel>>({
+  $set: { 'subInterfaceField.nestedObject': { a: 1, 'b': '2' } }
```
Comment (dariakp) [inline test/types/community/collection/updateX.test-d.ts:221]:
```suggestion
  $set: { 'subInterfaceField.nestedObject': { a: 1, b: '2' } }
```

**Review by dariakp — COMMENTED** [review]
lint fix

**Review by dariakp — APPROVED** [review]

**Review by nbbeeken — APPROVED** [review]

**Comment by coyotte508** [comment]
PS: Would it be possible after this and the previous PR #3259 to get a mention in [the release notes for 4.8](https://github.com/mongodb/node-mongodb-native/releases/tag/v4.8.0) and / or be added to https://github.com/mongodb/node-mongodb-native/blob/main/CONTRIBUTORS.md ?

All the best!

**Comment by dariakp** [comment]
@coyotte508 Apologies for the oversight on the release notes - we always want to give credit to the community for the contributions, so I'll get that fixed right away. The contributors file is currently not maintained, but we have plans to start generating it from github data.

**Comment by jaydenseric** [comment]
When is this going to be published? I can't use v4.8 it seems due to the problems this PR intends to fix:

Screen Shot 2022-07-23 at 3 55 38 pm

**Comment by dariakp** [comment]
@jaydenseric We're looking to put out a patch tomorrow!

**Comment by jer-sen** [comment]
@coyotte508 this kind of updates fail type check:

`{ $set: { 'aaa.$.bbb': "ccc" } }`

`{ $set: { 'aaa.$[].bbb': "ccc" } }`

`{ $set: { 'aaa.$[ddd].bbb': "ccc" } }` (`arrayFilters` option)

**Comment by coyotte508** [comment]
@jer-sen for me there are no TS errors, but it's more lax (allows invalid types).

**Comment by jer-sen** [comment]
@coyotte508 I also have an issue with variables in keys:

`{ $set: { [`aaa.${1 + 1}`]: "bbb" } }`

It seems that TS give to `{ [`aaa.${1 + 1}`]: "bbb" }` the type `Record<string, 'bbb'>`.

Any idea how to fix this without a forced cast?

**Comment by jer-sen** [comment]
> @jer-sen for me there are no TS errors, but it's more lax (allows invalid types).

I think it's because I use readonly arrays:

`const _: MatchKeysAndValues<{ aaa: { bbb: 'ccc' }[] }> = { 'aaa.$.bbb': 'ccc' };` is ok
`const _: MatchKeysAndValues<{ aaa: readonly { bbb: 'ccc' }[] }> = { 'aaa.$.bbb': 'ccc' };` is not ok

Could you improve your PR to handle this case?

**Comment by coyotte508** [comment]
I think https://github.com/mongodb/node-mongodb-native/blob/main/src/mongo_types.ts#L532 is the problem, you could change `[Key, ...NestedPaths<Type[Key]>]` to `[Key, ...NestedPaths<Type[Key]>] | [Key]` like https://github.com/mongodb/node-mongodb-native/blob/main/src/mongo_types.ts#L534 two lines below.

You can do the PR if this changes fixes your issue :), or open an issue for someone of the mongodb team on JIRA.

I'm going on holiday myself!

**Comment by jer-sen** [comment]
Thanks @coyotte508 but I am a bit lost...
@dariakp maybe you could have a look?

**Comment by jer-sen** [comment]
@coyotte508 @dariakp I have created:

- an issue in TS repo for variables in keys: https://github.com/microsoft/TypeScript/issues/50220
- a PR for complex nested names which fixes also issue with readonly arrays: https://github.com/mongodb/node-mongodb-native/pull/3349

**Pull request merged**